src/openrc/rc.c

```diff
@@ -349,22 +356,44 @@ static char *get_krunlevel(void)
 static void
 add_pid(pid_t pid)
 {
+	sigset_t sm, oldsm;
+
 	RC_PID *p = xmalloc(sizeof(*p));
 	p->pid = pid;
+
+	/* this list might be accessed inside the signal handler. so block
+	 * singnals to ensure it doesn't get into an inconsistent state.
+	 */
+	sigfillset(&sm);
+	sigprocmask(SIG_SETMASK, &sm, &oldsm);
```

Comment:
Since the list is only accessed inside the `SIGCHLD` handler, I think blocking just that might be okay instead of blocking everything.

Reply:
I think this is short enough that keeping code simple is better, but no preference here.